packages/vite/src/node/build.ts

```diff
@@ -1247,3 +1411,162 @@ function areSeparateFolders(a: string, b: string) {
+    buildApp: options.buildApp ?? defaultBuildApp,
+  }
+}
+
+export type ResolvedBuilderOptions = Required<BuilderOptions>
+
+export async function createBuilder(
+  inlineConfig: InlineConfig = {},
+): Promise<ViteBuilder> {
+  const config = await resolveConfigToBuild(inlineConfig)
+
+  const environments: Record<string, BuildEnvironment> = {}
+
+  const builder: ViteBuilder = {
+    environments,
+    config,
+    async buildApp() {
+      if (config.build.watch) {
+        throw new Error(
+          'Watch mode is not yet supported in viteBuilder.buildApp()',
+        )
+      }
+      return config.builder.buildApp(builder)
+    },
+    async build(environment: BuildEnvironment) {
+      return buildEnvironment(environment.config, environment)
+    },
+  }
+
+  for (const name of Object.keys(config.environments)) {
+    const environmentOptions = config.environments[name]
+    const createEnvironment =
+      environmentOptions.build?.createEnvironment ??
+      ((name: string, config: ResolvedConfig) =>
+        new BuildEnvironment(name, config))
+
+    // We need to resolve the config again so we can properly merge options
+    // and get a new set of plugins for each build environment. The ecosystem
+    // expects plugins to be run for the same environment once they are created
+    // and to process a single bundle at a time (contrary to dev mode where
+    // plugins are built to handle multiple environments concurrently).
+    let environmentConfig = config
+    if (!config.builder.sharedConfigBuild) {
+      const patchConfig = (resolved: ResolvedConfig) => {
+        // Until the ecosystem updates to use `environment.options.build` instead of `config.build`,
+        // we need to make override `config.build` for the current environment.
+        // We can deprecate `config.build` in ResolvedConfig and push everyone to upgrade, and later
+        // remove the default values that shouldn't be used at all once the config is resolved
+        ;(resolved.build as ResolvedBuildOptions) = {
+          ...resolved.environments[name].build,
+          lib: false,
+        }
+      }
+      const patchPlugins = (
+        rawPlugins: (Plugin | BoundedPluginConstructor)[],
+      ) => {
+        // Force opt-in shared plugins
+        const environmentPlugins = [...rawPlugins]
+        let validMixedPlugins = true
+        for (let i = 0; i < environmentPlugins.length; i++) {
+          const environmentPlugin = environmentPlugins[i]
+          if (typeof environmentPlugin === 'function') {
+            continue
+          }
+          const sharedPlugin = config.rawPlugins[i]
+          if (
+            config.builder.sharedPlugins ||
+            environmentPlugin.sharedDuringBuild
+          ) {
+            if (environmentPlugin.name !== sharedPlugin.name) {
+              validMixedPlugins = false
+              break
+            }
+            environmentPlugins[i] = sharedPlugin
+          }
```

Comment:
It should work in `vite@6.0.0-alpha.10` if you do:
```ts
function myPlugin(): PluginOption {
  // setup something in the scope
  let ctx = ...

  const plugin: BoundedPluginConstructor = (environment) => {
    // each environment sees different scope from each `myPlugin` call
    ctx;

    return [...]
  }

  plugin.sharedDuringBuild = true

  return plugin
}
```

It isn't ideal, but given that we want this flag to be only for backward compatibility and we would try to deprecate and remove it later, I think it is still not a blocker to go with the functional instead of the `create` hook design.

Reply:
`alpha.10` works, thanks! Yeah, assigning to function is not ideal, but I think it's alright. What I ended up is to have all plugins with `sharedDuringBuild` set just before return. It's also possible to splat `transformPlugins` and `extractPlugin` inside `environmentPlugin`'s `BoundedPlugin[]`.

```ts
return [
    ...transformPlugins, // Plugin[]
    extractPlugin,       // Plugin
    environmentPlugin,   // BoundedPluginConstructor
  ].map(
    (plugin) => {
      plugin.sharedDuringBuild = true;
      return plugin;
    },
  )
```

Reply:
If this is going to be a common pattern, we could expose an utility:
```js
sharedDuringBuild(plugins: PluginOption) => PluginOption
```
It feels a bit overkill for now, but it could clean up your code if you use it a lot 👍🏼